crates/steel-core/src/primitives/strings.rs

```diff
@@ -221,24 +221,46 @@ pub fn string_ci_equals(left: &SteelString, right: &SteelString) -> bool {
 
 #[function(name = "string-ref", constant = true)]
 pub fn string_ref(value: &SteelString, index: usize) -> Result<SteelVal> {
-    if index >= value.len() {
-        stop!(Generic => "string-ref: index out of bounds: index: {}, string length: {}", index, value);
-    }
+    let res = if index < value.len() {
+        value.chars().nth(index)
+    } else {
+        None
+    };
 
-    Ok(SteelVal::CharV(value.as_str().chars().nth(index).unwrap()))
+    if let Some(ch) = res {
+        Ok(SteelVal::CharV(ch))
+    } else {
+        stop!(Generic => "string-ref: index out of bounds: index: {}, string length: {}", index, value.len());
```

Comment:
I'm leaving `value.len()` as the "string length", even though that might come off weird in some scenarios:

```
λ > (string-ref "千葉市" 4)
error[E11]: Generic
  ┌─ :1:2
  │
1 │ (string-ref "千葉市" 4)
  │  ^^^^^^^^^^ string-ref: index out of bounds: index: 4, string length: 9
```

Reply:
This seems fine to me